### ISSUE TITLE
Resizing game window while displaying the Add-ons Manager causes layout issues

### ISSUE THREAD

**Issue by shikadilord** [wesnoth/wesnoth#1771]
Shrinking the game window (for example, from 1920x1021) while the Add-ons Manager is being displayed results in the dialog layout breaking, gaining a horizontal scrollbar and some text running out of space to be displayed, as per the following screenshot.

![image](https://user-images.githubusercontent.com/489895/26994761-04f963b8-4d37-11e7-9a1c-3500c2a38690.png)

Conversely, expanding the game window from an initial size that's small enough for the Descriptions panel to not be initially displayed doesn't result in the Descriptions panel being added to the layout.

![spectacle n26081](https://user-images.githubusercontent.com/489895/26994775-14c3b56e-4d37-11e7-8e1d-300ccbe16f65.png)

**Comment by gfgtdf** [comment]
This is mostlikeley becasue the code uses gui2 [resulution] that gets chosen when the dialog is created and not rechosen when the window is resized. The addon manager is not the only dialog that suffers from this probem, you can also observe this in the mp lobby.

**Comment by CelticMinstrel** [comment]
I thought the problem was fixed in the MP Lobby? I do recall it was fixed in the titlescreen... I suppose the MP Lobby may have design problems making a fix harder though.

**Comment by Vultraz** [comment]
Would it be possible to do something like add a hook to trigger `invalidate_layout` on resize? Would that help?

**Comment by CelticMinstrel** [comment]
No it would not, because invalidate_layout does not reload from the window WML. The only way currently to fix this is to actually close the dialog and re-open it.

Perhaps some facility could be added in modal_dialog to facilitate this, though.

**Comment by GregoryLundberg** [comment]
May be related to #1755. The initial paint was correct for the Normal sized window; but the window was Maximized and no layout or paint occurred for that size.

**Comment by Vultraz** [comment]
No, this is a different issue.

**Comment by gfgtdf** [comment]
> I thought the problem was fixed in the MP Lobby? I do recall it was fixed in the titlescreen... I suppose the MP Lobby may have design problems making a fix harder though.

Ye that inromation was quite old, I just tested again and now (1.13.8) wesnoth just crashses when i resize the window in the mp lobby :/ (`Caught general exception: std::exception`)

**Comment by gfgtdf** [comment]
currently changing the dialogs resolution type requires recreating the dialog/window object, if we really wanted to fix this issue we'd probably need to to save the dialogs state to an external object, recreate the dialog and then reapply the state. But i don't think we have the time to fix this one before 1.14 comes ou.

**Comment by CelticMinstrel** [comment]
I don't think it requires recreating the dialog object (only the window object). In other words, you can close the dialog, then re-show it. Of course you'd also want to restore the currently-viewed addon and such, too.

**Comment by gfgtdf** [comment]
while it is of course possible to refactor the dialog classes to make it possible, i don't think it's easier than recreating the dialog object aswell, often enough the dialog contains references to the window object or child objects of the window that need to be updated.

**Comment by CelticMinstrel** [comment]
It's true that dialogs often hold references to elements of the window, but those are assigned in the `pre_show` function, so they'd be reassigned if you re-show the same dialog instance. Basically all you should need is something like this:

```
gui2::addons_manager mgr;

while(!mgr.should_close()) {
    mgr.show();
}
```

Then make the dialog hook into resize events and close in a way that returns false from `should_close`.

**Comment by CelticMinstrel** [comment]
I don't currently plan to work on this, so unassigning for now. May take it up again later.

**Comment by Wedge009** [comment]
Checking add-on manager in 1.14.9, I see the first image is still an issue but I couldn't replicate the second one.

**Comment by jostephd** [comment]
> I couldn't replicate the second one.

I can (in master, didn't try 1.14). The key thing to know here is that the addon manager dialog has two different layouts depending on the window size. If I open wesnoth, resize window to 800x600 (using the preferences dialog), open addon manager, maximize wesnoth, I get the second screenshot.

**Comment by Wedge009** [comment]
Okay, yes, going from small to large I can replicate the second image too.
